providers/base/bin/snap_confinement_test.py

```diff
@@ -17,64 +17,173 @@
+            )
         )
-    )
-
-    missing_features = []
-    if "apparmor" not in sandbox_features:
-        logging.error("Cannot find 'apparmor' in sandbox-features")
-    else:
-        for feature in features_should_include:
-            if feature not in sandbox_features["apparmor"]:
-                missing_features.append(feature)
-
-    if missing_features:
-        logging.error("Cannot find '{}' in apparmor".format(missing_features))
-
-    categories_to_check = ["mount", "udev"]
-    for category in categories_to_check:
-        if category not in sandbox_features:
+
+        missing_features = []
+        if "apparmor" not in sandbox_features:
+            logging.error("Cannot find 'apparmor' in sandbox-features")
+        else:
+            for feature in self.features_should_include:
+                if feature not in sandbox_features["apparmor"]:
+                    missing_features.append(feature)
+
+        if missing_features:
             logging.error(
-                "Cannot find '{}' in sandbox-features".format(category)
+                "Cannot find '%s' in apparmor", missing_features
             )
-            break
-        for feature in sandbox_features[category]:
-            if "cgroup-v2" in feature:
-                logging.error("cgroup({}) must NOT be v2".format(feature))
 
-    return sandbox_features_output
+        categories_to_check = ["mount", "udev"]
+        for category in categories_to_check:
+            if category not in sandbox_features:
+                logging.error(
+                    "Cannot find '%s' in sandbox-features", category
+                )
+                break
+            for feature in sandbox_features[category]:
+                if "cgroup-v2" in feature:
+                    logging.error("cgroup(%s) must NOT be v2", feature)
+
+        return sandbox_features_output
+
+
+class SnapsConfinement:
+    """
+    Test the confinement status of all installed snaps.
+
+    Attributes:
+        whitelist_snaps (list): A list of snap names or regex patterns
+            that are exempted from the confinement check.
+    """
+
+    whitelist_snaps = [
+        "bugit",
+        r"checkbox.*",
+        "mir-test-tools",
+        "graphics-test-tools",
+    ]
+
+    def invoked(self):
+        """
+        Check the confinement of all snaps installed in the system.
+
+        A snap confinement should be 'strict', devmode should be False,
+        and should not have a sideloded revision starts with 'x'.
+
+        Returns:
+            int: Exit code. 0 if the test passes for all snaps,
+                 otherwise 1.
+        """
+        data = Snapd().list()
+        exit_code = 0
+        for snap in data:
+            snap_name = snap.get("name")
+            snap_confinement = snap.get("confinement")
+            snap_devmode = snap.get("devmode")
+            snap_revision = snap.get("revision")
+
+            if snap_name is None:
```

Comment:
I'm thinking that if we can't get the snap name, there's no point in proceeding with the other `if`s.

Reply:
You're right. I'll apply what you said in the next commit.